### PR TITLE
Milp encoding

### PR DESCRIPTION
**Pull request by guykatzz**

### CODE REVIEW

File: src/nlr/MILPFormulator.cpp [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+              region; we add the indicator variable a \in {0,1}:
+
+                y <= x - l ( 1 - a )
+                y <= u a
+
+                Or, alternatively:
+
+                y - x -la <= -l
+                y - ua <= 0
+            */
```
Comment (anwu1219) [inline src/nlr/MILPFormulator.cpp:288]: Would the constraints
y >= 0 and y <= x be captured?

File: src/nlr/LPFormulator.cpp [hunk 2]
```diff
@@ -61,93 +64,259 @@ double LPFormulator::solveLPRelaxation( const Map<unsigned, Layer *> &layers,
...
 
-            if ( ub < layer.second->getUb( i ) )
+    gurobiEnd = TimeUtils::sampleMicro();
+
+    LPFormulator_LOG( Stringf( "Number of tighter bounds found by Gurobi: %u. Sign changes: %u. Cutoffs: %u\n",
+                               tighterBoundCounter, signChanges, cutoffs ).ascii() );
+    LPFormulator_LOG( Stringf( "Seconds spent Gurobiing: %llu\n", TimeUtils::timePassed( gurobiStart, gurobiEnd ) / 1000000 ).ascii() );
+}
+
+void LPFormulator::optimizeBoundsWithLpRelaxation( const Map<unsigned, Layer *> &layers )
```
Comment (anwu1219) [inline src/nlr/LPFormulator.cpp:192]: Is this method called anywhere?

File: src/nlr/MILPFormulator.cpp (hunk 1 above)
Comment (guykatzz) [reply src/nlr/MILPFormulator.cpp:288]: Yes, these constraints are part of the underlying LP triangular constraints, which are added earlier by the lpFormulator

File: src/nlr/LPFormulator.cpp (hunk 2 above)
Comment (guykatzz) [reply src/nlr/LPFormulator.cpp:192]: Added a configuration option that controls with of the 4 methods gets called (LP/MILP x Inc/non-Inc)

File: src/nlr/LPFormulator.cpp [hunk 3]
```diff
@@ -61,93 +64,259 @@ double LPFormulator::solveLPRelaxation( const Map<unsigned, Layer *> &layers,
     return result;
 }
 
-void LPFormulator::optimizeBoundsWithLpRelaxation( const Map<unsigned, Layer *> &layers )
+void LPFormulator::optimizeBoundsWithIncrementalLpRelaxation( const Map<unsigned, Layer *> &layers )
```
[the remark this answers is not in the document]
Comment (guykatzz) [reply src/nlr/LPFormulator.cpp:67]: same as above

File: src/nlr/MILPFormulator.cpp [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+            // Maximize, using just the LP relaxation for the current layer
+            if ( tightenUpperBound( gurobi, layer, j, variable, currentUb, newUb ) )
+                continue;
+
+            // Minimize, using just the LP relaxation for the current layer
+            if ( tightenLowerBound( gurobi, layer, j, variable, currentLb, newLb ) )
+                continue;
+
+            // Add the MILP encoding
+            addLayerToModel( gurobi, layer );
```
[the remark this answers is not in the document]
Comment (guykatzz) [reply src/nlr/MILPFormulator.cpp]: Good point. I optimized this aspect of the code, please have a look.

[the remark this answers is not in the document]
Comment (anwu1219) [reply src/nlr/MILPFormulator.cpp]: Looks good to me.

**Pull request merged**